graph/builder.go

```diff
@@ -1761,53 +1761,108 @@ func (b *Builder) IsKnownEdge(chanID lnwire.ShortChannelID) bool {
+		disabled = update.IsDisabled()
+		isNode1  = update.IsNode1()
+	)
+
+	switch upd := update.(type) {
+	case *lnwire.ChannelUpdate1:
+		timestamp := time.Unix(int64(upd.Timestamp), 0)
+
+		edge1Timestamp, edge2Timestamp, exists, isZombie, err :=
+			b.cfg.Graph.HasChannelEdge1(chanID.ToUint64())
+		if err != nil {
+			log.Debugf("Check stale edge policy got error: %v", err)
+
+			return false
+		}
 
-	// If we know of the edge as a zombie, then we'll make some additional
-	// checks to determine if the new policy is fresh.
-	if isZombie {
-		// When running with AssumeChannelValid, we also prune channels
-		// if both of their edges are disabled. We'll mark the new
-		// policy as stale if it remains disabled.
-		if b.cfg.AssumeChannelValid {
-			isDisabled := flags&lnwire.ChanUpdateDisabled ==
-				lnwire.ChanUpdateDisabled
-			if isDisabled {
-				return true
+		// If we know of the edge as a zombie, then we'll make some
+		// additional checks to determine if the new policy is fresh.
+		if isZombie {
+			// When running with AssumeChannelValid, we also prune
+			// channels if both of their edges are disabled. We'll
+			// mark the new policy as stale if it remains disabled.
+			if b.cfg.AssumeChannelValid {
+				if disabled {
+					return true
+				}
 			}
+
+			// Otherwise, we'll fall back to our usual
+			// ChannelPruneExpiry.
+			return time.Since(timestamp) > b.cfg.ChannelPruneExpiry
 		}
 
-		// Otherwise, we'll fall back to our usual ChannelPruneExpiry.
-		return time.Since(timestamp) > b.cfg.ChannelPruneExpiry
-	}
+		// If we don't know of the edge, then it means it's fresh (thus
+		// not stale).
+		if !exists {
+			return false
+		}
 
-	// If we don't know of the edge, then it means it's fresh (thus not
-	// stale).
-	if !exists {
-		return false
-	}
+		// As edges are directional edge node has a unique policy for
+		// the direction of the edge they control. Therefore we first
+		// check if we already have the most up to date information for
+		// that edge. If so, then we can exit early.
+		switch {
+		case isNode1:
+			return !edge1Timestamp.Before(timestamp)
 
-	// As edges are directional edge node has a unique policy for the
-	// direction of the edge they control. Therefore, we first check if we
-	// already have the most up-to-date information for that edge. If so,
-	// then we can exit early.
-	switch {
-	// A flag set of 0 indicates this is an announcement for the "first"
-	// node in the channel.
-	case flags&lnwire.ChanUpdateDirection == 0:
-		return !edge1Timestamp.Before(timestamp)
-
-	// Similarly, a flag set of 1 indicates this is an announcement for the
-	// "second" node in the channel.
-	case flags&lnwire.ChanUpdateDirection == 1:
-		return !edge2Timestamp.Before(timestamp)
+		case !isNode1:
+			return !edge2Timestamp.Before(timestamp)
+		}
+
+	case *lnwire.ChannelUpdate2:
+		height := upd.BlockHeight
+
+		edge1Height, edge2Height, exists, isZombie, err :=
+			b.cfg.Graph.HasChannelEdge2(chanID.ToUint64())
```

Comment:
I think if we can come up with an abstraction over the return values here (height vs timestamp), then we can condense this function a bit and eliminate some of the duplication.